src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -434,15 +434,23 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                     break;
 
                 case 3:
-                    assert(isRMW);
-                    if (targetReg != op1Reg)
+                    if (isRMW)
                     {
-                        assert(targetReg != op2Reg);
-                        assert(targetReg != op3Reg);
+                        if (targetReg != op1Reg)
+                        {
+                            assert(targetReg != op2Reg);
+                            assert(targetReg != op3Reg);
 
-                        GetEmitter()->emitIns_Mov(INS_mov, emitTypeSize(node), targetReg, op1Reg, /* canSkip */ true);
+                            GetEmitter()->emitIns_Mov(INS_mov, emitTypeSize(node), targetReg, op1Reg,
+                                                      /* canSkip */ true);
+                        }
+                        GetEmitter()->emitIns_R_R_R(ins, emitSize, targetReg, op2Reg, op3Reg, opt);
+                    }
+                    else
+                    {
+                        GetEmitter()->emitIns_R_R_R_R(ins, emitSize, targetReg, op1Reg, op2Reg, op3Reg,
+                                                    opt, INS_SCALABLE_OPTS_UNPREDICATED);
```

Comment:
I thought that `SEL` was always predicated so passing in `OPTS_UNPREDICATED` seems incorrect, no?

https://developer.arm.com/documentation/ddi0602/2024-03/SVE-Instructions/SEL--vectors---Conditionally-select-elements-from-two-vectors-?lang=en

Reply:
`INS_SCALABLE_OPTS_UNPREDICATED` is temporary and will be removable once predicate register allocation is complete.  There's a few places it's currently used awkwardly.

Reply:
What's required to get it to be used correctly?

I would have thought it was strictly incorrect here given this is a predicated instruction.

Reply:
emitIns_R_R_R_R with no sopt:
`sel <Pd>.B, <Pg>, <Pn>.B, <Pm>.B`
https://docsmirror.github.io/A64/2023-09/sel_z_p_zz.html

emitIns_R_R_R_R with INS_SCALABLE_OPTS_UNPREDICATED:
`sel <Zd>.<T>, <Pv>, <Zn>.<T>, <Zm>.<T>`
https://docsmirror.github.io/A64/2023-09/sel_p_p_pp.html

Here it's using `UNPREDICATED` to mean args 1, 3 and 4 are not predicates. Which is not how the enum was intended for use (it was meant to be used for instructions which the result is not masked)

Fix would be to add a new enum entry and find all the places it's not correctly used. But there are no clashes and it's being removed soon anyway.

Reply:
> But there are no clashes and it's being removed soon anyway.

+1 on that. I will keep it as-is for now. There is already a `TODO` around those enums that they will be removed.